src/libraries/Common/tests/System/Security/Cryptography/AlgorithmImplementations/SlhDsa/SlhDsaContractTests.cs

```diff
@@ -0,0 +1,221 @@
+            slhDsa.AddFillDestination(1);
+
+            slhDsa.SignData(testData, signature.AsSpan(PaddingSize, signatureSize + extraBytes), testContext);
+            Assert.Equal(1, slhDsa.SignDataCoreCallCount);
+            AssertExpectedFill(signature, fillElement: 1, paddingElement: 42, PaddingSize, signatureSize);
+        }
+
+        [Theory]
+        [MemberData(nameof(SlhDsaTestData.AlgorithmsData), MemberType = typeof(SlhDsaTestData))]
+        public static void VerifyData_CallsCore(SlhDsaAlgorithm algorithm)
+        {
+            using SlhDsaMockImplementation slhDsa = SlhDsaMockImplementation.Create(algorithm);
+
+            int signatureSize = algorithm.SignatureSizeInBytes;
+            byte[] testSignature = CreatePaddedFilledArray(signatureSize, 42);
+            byte[] testData = [2];
+            byte[] testContext = [3];
+            bool returnValue = false;
+
+            slhDsa.VerifyDataCoreHook = (_, _, _) => returnValue;
+            slhDsa.AddDataBufferIsSameAssertion(testData);
+            slhDsa.AddContextBufferIsSameAssertion(testContext);
+            slhDsa.AddSignatureBufferIsSameAssertion(testSignature.AsMemory(PaddingSize, signatureSize));
+
+            // Since `returnValue` is true, this shows the Core method doesn't get called for the wrong sized signature.
+            returnValue = true;
+            AssertExtensions.FalseExpression(slhDsa.VerifyData(testData, testSignature.AsSpan(PaddingSize, signatureSize - 1), testContext));
+            Assert.Equal(0, slhDsa.VerifyDataCoreCallCount);
+
+            AssertExtensions.FalseExpression(slhDsa.VerifyData(testData, testSignature.AsSpan(PaddingSize, signatureSize + 1), testContext));
+            Assert.Equal(0, slhDsa.VerifyDataCoreCallCount);
+
+            // But does for the right one.
+            AssertExtensions.TrueExpression(slhDsa.VerifyData(testData, testSignature.AsSpan(PaddingSize, signatureSize), testContext));
+            Assert.Equal(1, slhDsa.VerifyDataCoreCallCount);
+
+            // And just to prove that the Core method controls the answer...
+            returnValue = false;
+            AssertExtensions.FalseExpression(slhDsa.VerifyData(testData, testSignature.AsSpan(PaddingSize, signatureSize), testContext));
+            Assert.Equal(2, slhDsa.VerifyDataCoreCallCount);
+        }
+
+        [Theory]
+        [MemberData(nameof(SlhDsaTestData.AlgorithmsData), MemberType = typeof(SlhDsaTestData))]
+        public static void Dispose_CallsVirtual(SlhDsaAlgorithm algorithm)
+        {
+            SlhDsaMockImplementation slhDsa = SlhDsaMockImplementation.Create(algorithm);
+            bool disposeCalled = false;
+
+            // First Dispose call should invoke overridden Dispose should be called
+            slhDsa.DisposeHook = (bool disposing) =>
+            {
+                AssertExtensions.TrueExpression(disposing);
+                disposeCalled = true;
```

Comment:
Would this test be more powerful if instead of bool it was int/++ and it checked that it went from 0 to 1 to (staying) 1 across the multiple calls to Dispose?

Reply:
Nevermind, I see that it did that by replacing the hook with failure after the fist call.